File: frontend/src/types/weather.ts
export interface StationInfo {
  id: string;
  name: string;
  location: string;
  latitude: number;
  longitude: number;
  grid_wfo: string;
  grid_x: number;
  grid_y: number;
  elevation_ft?: number;
}

export interface CurrentConditions {
  station_id: string;
  station_name: string;
  location: string;
  observation_time: string;
  data_retrieved_time: string;
  temperature_f: number;
  temperature_c: number;
  feels_like_f: number;
  dewpoint_f?: number;
  relative_humidity_pct?: number;
  wind_speed_mph?: number;
  wind_direction_deg?: number;
  wind_direction_cardinal?: string;
  wind_gust_mph?: number;
  pressure_inhg?: number;
  pressure_hpa?: number;
  precipitation_last_hour_in?: number;
  weather_condition: string;
  weather_icon?: string;
  visibility_miles?: number;
  data_source: string;
}

export interface HourlyForecastItem {
  timestamp: string;
  formatted_time: string;
  temperature_f: number;
  dewpoint_f?: number;
  relative_humidity_pct?: number;
  precipitation_probability_pct: number;
  wind_speed_mph: number;
  wind_direction_cardinal: string;
  wind_gust_mph?: number;
  weather_condition: string;
  weather_icon?: string;
  is_daytime: boolean;
}

export interface DailyForecastItem {
  date: string;
  day_name: string;
  min_temp_f: number;
  max_temp_f: number;
  avg_temp_f: number;
  hdd: number;
  precipitation_probability_pct: number;
  snow_probability_pct: number;
  wind_speed_mph: number;
  wind_direction_cardinal: string;
  weather_condition: string;
  weather_icon?: string;
  detailed_forecast?: string;
}

export interface WeatherAlert {
  id: string;
  event: string;
  severity: string;
  urgency: string;
  certainty: string;
  area_desc: string;
  effective_time: string;
  expires_time: string;
  headline: string;
  description: string;
  instruction?: string;
  sender_name: string;
  is_operational_banner: boolean;
  source: string;
}

export interface GasIndicators {
  station_id: string;
  station_name: string;
  hdd_base_temp: number;
  daily_hdd: number;
  forecast_hdd_today: number;
  cumulative_hdd_3d: number;
  cumulative_hdd_5d: number;
  cumulative_hdd_7d: number;
  temp_change_24h: number;
  expected_temp_change_next_24h: number;
  trend_3d: string;
  trend_5d: string;
  cold_weather_level: 'Normal' | 'Elevated' | 'High' | 'Extreme';
  cold_weather_label: string;
  rapid_drop_detected: boolean;
  rapid_drop_magnitude: number;
  rapid_drop_message?: string;
  composite_weather_index: number;
  composite_weather_index_label: string;
  calculation_note: string;
}

export interface HistoricalDataPoint {
  date: string;
  avg_temp_f: number;
  min_temp_f: number;
  max_temp_f: number;
  hdd: number;
  precipitation_in: number;
  snowfall_in: number;
}

export interface HistoricalVsForecast {
  recent_7d_avg_temp: number;
  forecast_7d_avg_temp: number;
  temp_difference: number;
  recent_7d_total_hdd: number;
  forecast_7d_total_hdd: number;
  hdd_difference: number;
  recent_min_temp: number;
  forecast_min_temp: number;
  recent_max_temp: number;
  forecast_max_temp: number;
}

export interface StationComparisonData {
  timestamp: string;
  temperature_delta: number;
  hdd_delta_7d: number;
  summary: string;
  central_park: {
    station: StationInfo;
    current: CurrentConditions;
    indicators: GasIndicators;
    daily: DailyForecastItem[];
    alerts: WeatherAlert[];
  };
  white_plains: {
    station: StationInfo;
    current: CurrentConditions;
    indicators: GasIndicators;
    daily: DailyForecastItem[];
    alerts: WeatherAlert[];
  };
}

export interface UserSettings {
  hdd_base_temp: number;
  cold_threshold_elevated: number;
  cold_threshold_high: number;
  cold_threshold_extreme: number;
  rapid_drop_threshold: number;
  auto_refresh_minutes: number;
  temp_unit: string;
}

File: backend/app/models/weather.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

class StationInfo(BaseModel):
    id: str = Field(..., description="Station ICAO or identifier, e.g., KNYC, KHPN")
    name: str = Field(..., description="Official station name")
    location: str = Field(..., description="Geographic locality, e.g. Central Park, NY")
    latitude: float
    longitude: float
    grid_wfo: str = Field("OKX", description="NWS Weather Forecast Office")
    grid_x: int
    grid_y: int
    elevation_ft: Optional[float] = None
    latest_observation_time: Optional[datetime] = None
    is_active: bool = True

class CurrentConditions(BaseModel):
    station_id: str
    station_name: str
    location: str
    observation_time: datetime
    data_retrieved_time: datetime
    temperature_f: float
    temperature_c: float
    feels_like_f: float
    dewpoint_f: Optional[float] = None
    relative_humidity_pct: Optional[float] = None
    wind_speed_mph: Optional[float] = None
    wind_direction_deg: Optional[float] = None
    wind_direction_cardinal: Optional[str] = None
    wind_gust_mph: Optional[float] = None
    pressure_inhg: Optional[float] = None
    pressure_hpa: Optional[float] = None
    precipitation_last_hour_in: Optional[float] = 0.0
    weather_condition: str = "Unknown"
    weather_icon: Optional[str] = None
    visibility_miles: Optional[float] = None
    data_source: str = "NOAA / National Weather Service (api.weather.gov)"

class HourlyForecastItem(BaseModel):
    timestamp: datetime
    formatted_time: str
    temperature_f: float
    dewpoint_f: Optional[float] = None
    relative_humidity_pct: Optional[float] = None
    precipitation_probability_pct: Optional[float] = 0.0
    wind_speed_mph: Optional[float] = 0.0
    wind_direction_cardinal: Optional[str] = "N"
    wind_gust_mph: Optional[float] = None
    weather_condition: str
    weather_icon: Optional[str] = None
    is_daytime: bool = True

class DailyForecastItem(BaseModel):
    date: str
    day_name: str
    min_temp_f: float
    max_temp_f: float
    avg_temp_f: float
    hdd: float
    precipitation_probability_pct: float = 0.0
    snow_probability_pct: float = 0.0
    wind_speed_mph: float = 0.0
    wind_direction_cardinal: str = "N"
    weather_condition: str
    weather_icon: Optional[str] = None
    detailed_forecast: Optional[str] = None

class WeatherAlert(BaseModel):
    id: str
    event: str
    severity: str  # Extreme, Severe, Moderate, Minor, Unknown
    urgency: str
    certainty: str
    area_desc: str
    effective_time: datetime
    expires_time: datetime
    headline: str
    description: str
    instruction: Optional[str] = None
    sender_name: str
    is_operational_banner: bool = False
    source: str = "NOAA / National Weather Service"

class GasIndicators(BaseModel):
    station_id: str
    station_name: str
    hdd_base_temp: float = 62.0
    daily_hdd: float
    forecast_hdd_today: float
    cumulative_hdd_3d: float
    cumulative_hdd_5d: float
    cumulative_hdd_7d: float
    temp_change_24h: float
    expected_temp_change_next_24h: float
    trend_3d: str  # Warming, Stable, Cooling
    trend_5d: str  # Warming, Stable, Cooling
    cold_weather_level: str  # Normal, Elevated, High, Extreme
    cold_weather_label: str
    rapid_drop_detected: bool
    rapid_drop_magnitude: float
    rapid_drop_message: Optional[str] = None
    composite_weather_index: float
    composite_weather_index_label: str
    calculation_note: str = "Con Edison Gas Control: Actual Heating Degree Days (AHDD) = 62°F - Daily Mean Temperature; does not represent metered gas sendout."

class HistoricalDataPoint(BaseModel):
    date: str
    avg_temp_f: float
    min_temp_f: float
    max_temp_f: float
    hdd: float
    precipitation_in: float
    snowfall_in: float = 0.0

class HistoricalVsForecast(BaseModel):
    recent_7d_avg_temp: float
    forecast_7d_avg_temp: float
    temp_difference: float
    recent_7d_total_hdd: float
    forecast_7d_total_hdd: float
    hdd_difference: float
    recent_min_temp: float
    forecast_min_temp: float
    recent_max_temp: float
    forecast_max_temp: float

class StationComparisonData(BaseModel):
    timestamp: datetime
    central_park: Dict[str, Any]
    white_plains: Dict[str, Any]
    temperature_delta: float
    hdd_delta_7d: float
    summary: str

class UserSettings(BaseModel):
    hdd_base_temp: float = 62.0
    cold_threshold_elevated: float = 34.0
    cold_threshold_high: float = 24.0
    cold_threshold_extreme: float = 14.0
    rapid_drop_threshold: float = 15.0
    auto_refresh_minutes: int = 30
    temp_unit: str = "F"
